### PR TITLE
Search user

### PR DESCRIPTION
**Pull request by wilderino**
The search can now search for user and display the result in similarity order

### CODE REVIEW

File: app/src/main/java/ch/epfl/polychef/users/SearchUser.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        return INSTANCE;
+    }
+
+    public void searchForUser(String query, CallHandler<List<Miniatures>> caller) {
+        searchUser(query, this::compareSimilarity, caller);
+    }
+
+    private void searchUser(String query, BiFunction<String, User, Boolean> comparator, CallHandler<List<Miniatures>> caller) {
+        DatabaseReference nameRef = getDatabase().getReference(UserStorage.DB_NAME);
+        nameRef.addValueEventListener(new ValueEventListener() {
```
Comment (yanickpauloamaro) [inline app/src/main/java/ch/epfl/polychef/users/SearchUser.java]: Do we really need `addValueEventListener` here? Wouldn't a `addListenerForSingleValueEvent` be enough?

File: app/src/androidTest/java/ch/epfl/polychef/images/ImageHandlerTest.java [hunk 2]
```diff
@@ -41,13 +41,17 @@
...
+    ImageStorage mockImageStorage;
+
     ImageHandler realImageHandler;
     ImageHandler fakeImageHandler;
 
     @Before
     public void setImagesAndMock() {
         MockitoAnnotations.initMocks(this);
         when(task.addOnSuccessListener(any())).thenReturn(task);
+        when(mockImageStorage.upload(any(), any(), any(), any())).thenReturn(task);
```
Comment (simhayoz) [inline app/src/androidTest/java/ch/epfl/polychef/images/ImageHandlerTest.java]: Don't use raw `any()` instead specify the class of the element (i.e. `.upload(any(byte[].class), anyString(), anyString(), anyString())`)

File: app/src/androidTest/java/ch/epfl/polychef/images/ImageStorageTest.java [hunk 3]
```diff
@@ -46,9 +52,11 @@ public void initMockFirebaseStorage() {
         MockitoAnnotations.initMocks(this);
         successTask = Tasks.forResult(new byte[] {1, 2, 3});
         failTask = Tasks.forException(new Exception());
+        imageStorageSpy = Mockito.spy(ImageStorage.getInstance());
+        when(imageStorageSpy.getStorage()).thenReturn(firebaseStorage);
```
Comment (simhayoz) [inline app/src/androidTest/java/ch/epfl/polychef/images/ImageStorageTest.java:56]: Nice way to mock singleton pattern classes, we may use this elsewhere as well

**Review by simhayoz — CHANGES_REQUESTED** [review]
Very good (and enormous) job!
For the `RecipeStorage`, why not keep the List<Recipe> to send back? Is it a problem of type inferring with the `List<Recipe>` - `List<User>`?

For later:
As we don't have much classes that implements `Miniatures`, we could imagine doing something like:
```java
public interface Miniatures {
    public boolean isRecipe();
    public boolean isUser();
    public Recipe asRecipe();
    public User asUser();
}
```
and just implements these method in `Recipe` and `User`
e.g. for  `Recipe`:
```java
    /*...*/
    public boolean isRecipe() {
        return true;
    }
    public boolean isUser() {
        return false;
    }
    public Recipe asRecipe() {
        return this;
    }
    public User asUser() {
        throw new IllegalAssignementException(); // Or whatever exception
    }
```
Doing this will prevent the `miniaturesList.get(position).getClass().equals(Recipe.class)` and change it to: `miniaturesList.get(position).isRecipe()`

File: app/src/main/java/ch/epfl/polychef/utils/Similarity.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
+package ch.epfl.polychef.utils;
+
+public class Similarity {
```
Comment (Ingloryon) [inline app/src/main/java/ch/epfl/polychef/utils/Similarity.java:3]: J'allais demander si il y avait des methodes deja built in en java mais je suis tombé sur le post que vous avez du utiliser:
https://stackoverflow.com/questions/955110/similarity-string-comparison-in-java

Vous avez choisi la deuxieme approche parce que c'etait plus simple à implémenter / à tester ?

File: app/src/androidTest/java/ch/epfl/polychef/images/ImageStorageTest.java (hunk 3 above)
Comment (yanickpauloamaro) [reply app/src/androidTest/java/ch/epfl/polychef/images/ImageStorageTest.java:56]: Yeah, I wish we had found it sooner

File: app/src/androidTest/java/ch/epfl/polychef/images/ImageHandlerTest.java (hunk 2 above)
Comment (yanickpauloamaro) [reply app/src/androidTest/java/ch/epfl/polychef/images/ImageHandlerTest.java]: You are right, I'll change that

**Review by simhayoz — APPROVED** [review]

**Review by Ingloryon — APPROVED** [review]
Indeed there's quite some work ! Impressive job

**Pull request merged**